text/0113-follower-read-cache.md

```diff
@@ -0,0 +1,83 @@
+# Cache read index resposnses
+
+## Summary
+The purpose of this design is to reduce the CPU uitlization on the leader by caching the response of read index requests from followers. By avoiding the need to send these requests when there are no pending writes on the leader at a specific read timestamp, we can significantly reduce the CPU overhead caused by expensive read index RPCs.
+
+## Motivation
+
+In a cloud environment, customers often prefer to perform follower reads within the same AZ to avoid incurring cross-AZ network costs. However, each follower read requires sending a read index request to the leader to check for any pending write locks in memory. These RPC calls can be quite expensive and can significantly increase CPU utilization, especially in read-heavy traffic scenarios. This increase in CPU cost eliminates the cost savings achieved by avoiding cross-AZ traffic.
+To mitigate this issue in read-heavy environments, it is not necessary to send a read index request for every single follower read request. This is because, in 99% of cases, there are no pending locks. 
+
+## Detailed design
+
+We introduce a new concept last-read-index-ts. last-read-index-ts is updated with the read-ts in the follower under two specific conditions:
+1. There is no memory lock in that region.
+2. The apply_index in the follower is greater than the commit_index of the leader.
+In this proposal, response of read index messsages can be used to advance last-read-index-ts in follower. Ref counters per region need to be added in the leader to keep track of total number of pending locks and lock status need to be included in a raft read index response. These are the detailed steps on how a follower read works
+- Follower reads will initially compare the last-read-index-ts with the timestamp in the request, and only if the request timestamp is greater than the last-read-index-ts, a read index request will be sent.
+- Read index request on leader check the lock reference counter and send its status through raft message in read index response
+- read index response update last-read-index-ts if there are no locks at this timestamp in leader.
+
+This feature would be enabled by default and no new configuration will be added. 
+
+### Protocol
+
+A new field `memory_lock` will be added to `ReadIndexContext` to send the lock status to replica.
+
+```
+pub struct ReadIndexContext {
+    pub id: Uuid,
+    pub request: Option<raft_cmdpb::ReadIndexRequest>,
+    pub locked: Option<LockInfo>,
+    pub memory_lock: Option<u64>,
+}
+```
+### Code changes
+
```

Comment:
Better to describe how the `last-read-index-ts` is maintained in the code change part, like where it's stored and how it's read and modified.

Reply:
it is stored as part of read_progress. its modification is shown in line 50. I have added a structure also now.